src/include/btree.i

```diff
@@ -530,6 +530,9 @@ __wt_page_modify_set(WT_SESSION_IMPL *session, WT_PAGE *page)
 		WT_FULL_BARRIER();
 	}
 
+	if (S2C(session)->modified == 0)
+		S2C(session)->modified = 1;
+
```

Comment:
> Or is there some case where the btree is dirty on entry but the connection could be clean?

Yes, the intention is to make sure we update connection to dirty in the case btree is marked dirty and connection is not while entering the function. I don't have a concrete example on how this might happen. Consider the duration between marking connection clean after taking transaction id and marking btree clean after checkpointing this specific btree. During this time period btree is marked dirty and connection clean for any thread calling __wt_page_modify_set() in parallel to the checkpoint thread.

I think it is cleaner to not assume a coupling this tight between btree and connection modified states. We should make sure that __wt_page_modify_set() marks connection dirty if btree is supposed to be dirty, irrespective of what state these flags were when starting to execute the function.